pkg/webhook/server.go

```diff
@@ -169,51 +177,54 @@ func (s *Server) Start(ctx context.Context) error {
 	baseHookLog := log.WithName("webhooks")
 	baseHookLog.Info("Starting webhook server")
 
-	certPath := filepath.Join(s.CertDir, s.CertName)
-	keyPath := filepath.Join(s.CertDir, s.KeyName)
-
-	certWatcher, err := certwatcher.New(certPath, keyPath)
-	if err != nil {
-		return err
-	}
-
-	go func() {
-		if err := certWatcher.Start(ctx); err != nil {
-			log.Error(err, "certificate watcher error")
-		}
-	}()
-
 	tlsMinVersion, err := tlsVersion(s.TLSMinVersion)
 	if err != nil {
 		return err
 	}
 
 	cfg := &tls.Config{ //nolint:gosec
-		NextProtos:     []string{"h2"},
-		GetCertificate: certWatcher.GetCertificate,
-		MinVersion:     tlsMinVersion,
+		NextProtos: []string{"h2"},
+		MinVersion: tlsMinVersion,
 	}
+	// fallback TLS config ready, will now mutate if passer wants full control over it
+	for _, op := range s.TLSOpts {
+		op(cfg)
+	}
+
+	if cfg.GetCertificate == nil {
+		certPath := filepath.Join(s.CertDir, s.CertName)
+		keyPath := filepath.Join(s.CertDir, s.KeyName)
 
-	// load CA to verify client certificate
-	if s.ClientCAName != "" {
-		certPool := x509.NewCertPool()
-		clientCABytes, err := os.ReadFile(filepath.Join(s.CertDir, s.ClientCAName))
+		// Create the certificate watcher and
+		// set the config's GetCertificate on the TLSConfig
+		certWatcher, err := certwatcher.New(certPath, keyPath)
 		if err != nil {
-			return fmt.Errorf("failed to read client CA cert: %w", err)
+			return err
 		}
-
-		ok := certPool.AppendCertsFromPEM(clientCABytes)
-		if !ok {
-			return fmt.Errorf("failed to append client CA cert to CA pool")
+		cfg.GetCertificate = certWatcher.GetCertificate
+
+		go func() {
+			if err := certWatcher.Start(ctx); err != nil {
+				log.Error(err, "certificate watcher error")
+			}
+		}()
+
+		// load CA to verify client certificate
+		if s.ClientCAName != "" {
```

Comment:
🤔 Problem is that we're using CertDir and CertCAName to get the certificate from the local file

Reply:
We could also decouple by deprecating ClientCAName and introducing ClientCAFile

Then it's only a bit hacky until we remove ClientCAName

Reply:
Folks can configure CA validation through TLS options if they wish to do so outside the CertDir. The current comments seem fine to me in those cases, wdyt? cc @alvaroaleman

Reply:
Just feels strange to me to disable the client authentication when GetCertificate (which is only for the server cert) is set.

Reply:
The following would feel more logical:
* Use GetCertifiate if it is set
* Use ClientCAs and ClientAuth if one of them (?) is set

Otherwise fallback to using the configured fields

Reply:
Maybe add some validation that errors if any of the fields + `GetCertificate` is set?

Reply:
Fine for me!

Makes it a bit less implicit / more obvious

Reply:
PTAL, opted to both return errors, and allow to use ClientCAName w/ GetCertificate if folks still want

Reply:
Looks good, mind adding a testcase for `GetCertificate` not getting overriden if present?

Reply:
lgtm as well